pkg/util/vsphere/watcher/watcher.go

```diff
@@ -0,0 +1,674 @@
+					{
+						ObjectSet: []vimtypes.ObjectSpec{
+							{
+								Obj: obj,
+							},
+						},
+						PropSet: []vimtypes.PropertySpec{
+							{
+								Type: "VirtualMachine",
+								PathSet: []string{
+									extraConfigNamespaceNamePropPath,
+								},
+							},
+						},
+					},
+				},
+			})
+		if err != nil {
+			return err
+		}
+		namespace, name = namespacedNameFromObjContent(rep.Returnval)
+	}
+
+	if namespace != "" && name != "" {
+		r := Result{
+			Namespace: namespace,
+			Name:      name,
+			Ref:       obj,
+			Verified:  verified,
+		}
+
+		logger.Info("sending result", "result", r)
+
+		go func(r Result) {
+			w.chanResult <- r
+		}(r)
+	}
+
+	return nil
+}
+
+func checkExtraConfig(
+	aov vimtypes.ArrayOfOptionValue,
+	ignoredKeys map[string]struct{}) (ignore bool, namespace, name string) {
+
+	hasNonIgnoredKey := false
+
+	for j := range aov.OptionValue {
+		if ov := aov.OptionValue[j].GetOptionValue(); ov != nil {
+			// Get the namespace and name of the VM from the changes
+			// if they are present there.
+			if ov.Key == extraConfigNamespacedNameKey {
+				if namespace == "" || name == "" {
+					if s, ok := ov.Value.(string); ok {
+						namespace, name = namespacedNameFromString(s)
+					}
+				}
+			}
+			// Note if the key cannot be ignored.
+			if _, ok := ignoredKeys[ov.Key]; !ok {
+				hasNonIgnoredKey = true
+			}
+			if hasNonIgnoredKey && (namespace != "" && name != "") {
+				break
+			}
+		}
+	}
+
+	return !hasNonIgnoredKey, namespace, name
+}
+
+func (w *Watcher) add(ctx context.Context, ref moRef, id string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.pc == nil {
+		// The property collector is nil means the watcher is closed/stopped;
+		// return early.
+		return nil
+	}
+
+	if _, ok := w.cv[ref]; ok {
+		w.cvr[ref][id] = struct{}{}
+		return nil
+	}
+
+	cv, err := w.vm.CreateContainerView(
+		ctx,
+		ref,
+		[]string{virtualMachineType},
+		true)
+	if err != nil {
+		return err
+	}
+
+	if _, err := w.lv.Add(
+		ctx,
+		[]vimtypes.ManagedObjectReference{cv.Reference()}); err != nil {
+
+		if err2 := cv.Destroy(context.Background()); err2 != nil {
+			return fmt.Errorf(
+				"failed to destroy container view after adding "+
+					"it to list failed: addErr=%w, destroyErr=%w", err, err2)
+		}
+
+		return err
+	}
+
+	w.cv[ref] = cv
+	if w.cvr[ref] == nil {
+		w.cvr[ref] = map[string]struct{}{}
+	}
+	w.cvr[ref][id] = struct{}{}
+
+	return nil
+}
+
+func (w *Watcher) remove(_ context.Context, ref moRef, id string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.pc == nil {
+		// The property collector is nil means the watcher is closed/stopped;
+		// return early.
+		return nil
+	}
+
+	cv, ok := w.cv[ref]
+	if !ok {
+		return nil
+	}
+
+	// Only remove the container from the list view if it has a ref count of
+	// one.
+	if len(w.cvr[ref]) > 1 {
+		delete(w.cvr[ref], id)
+		return nil
+	}
+
+	_, err := w.lv.Remove(context.Background(), []moRef{cv.Reference()})
```

Comment:
Why not use the Context param passed to the function instead of `context.Background()`?

Reply:
Because the reason the earlier one failed could be due to the context being cancelled.

Reply:
The reason the param has a context is in case I use a logger.